src/python/pants/engine/native.py

```diff
@@ -682,19 +682,26 @@ def __init__(self, visualize_to_dir):
   def visualize_to_dir(self):
     return self._visualize_to_dir
 
+  class BinaryLocationError(Exception): pass
+
   @memoized_property
   def binary(self):
     """Load and return the path to the native engine binary."""
     lib_name = '{}.so'.format(NATIVE_ENGINE_MODULE)
     lib_path = os.path.join(safe_mkdtemp(), lib_name)
-    with closing(pkg_resources.resource_stream(__name__, lib_name)) as input_fp:
-      # NB: The header stripping code here must be coordinated with header insertion code in
-      #     build-support/bin/native/bootstrap_code.sh
-      engine_version = input_fp.readline().decode('utf-8').strip()
-      repo_version = input_fp.readline().decode('utf-8').strip()
-      logger.debug('using {} built at {}'.format(engine_version, repo_version))
-      with open(lib_path, 'wb') as output_fp:
-        output_fp.write(input_fp.read())
+    try:
+      with closing(pkg_resources.resource_stream(__name__, lib_name)) as input_fp:
+        # NB: The header stripping code here must be coordinated with header insertion code in
+        #     build-support/bin/native/bootstrap_code.sh
+        engine_version = input_fp.readline().decode('utf-8').strip()
+        repo_version = input_fp.readline().decode('utf-8').strip()
+        logger.debug('using {} built at {}'.format(engine_version, repo_version))
+        with open(lib_path, 'wb') as output_fp:
+          output_fp.write(input_fp.read())
+    except (IOError, OSError) as e:
+      raise self.BinaryLocationError(
+        "Error loading the native engine binary from path {}: {}".format(lib_path, e),
```

Comment:
"loading from" is possibly a little misleading here; what it's actually doing is unpacking the native engine binary _to_ that path...